### PR TITLE
cpu/nrf52: add periph_uart_non_blocking to nrf52840

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

This PR adds `FEATURES_PROVIDED += periph_uart_nonblocking` for `nrf52840` devices. UARTE is not used to its full potential, since I could instead of writing single bytes use the tsrb buffer as the RAM buffer. But I found the logic much more complicated in this case since it required either:

- duplicated tx_buffers
- if using the tsrb buffer as the RAM buffer then it needs to keep track of where in the buffer we are
    - but then sent bytes from the tsrb buffer need to be poped once they are sent out so either:
      - at the END of transmission using `EVENT_ENDTX` and `TX.MAXCNT`
      - every time a byte is sent out using ` EVENT_TXDRDY`
        - but ` EVENT_TXDRDY` can overlap or be missed so at the end of transmission still need to POP the remaining ones
        - so required keeping track of the ones that were popped already, something like a `txrdy_cnt`

And then there is the case were trying to send with masked ISR. All in all the code looked quite complex when I wrote it (I had chosen the second option), but I can PR it as an alternative.

### Testing procedure

`BOARD=nrf52840-mdk make -C tests/periph_uart_nonblocking/ flash test -j3`

```
main(): This is RIOT! (Version: 2021.04-devel-725-gcca249-pr_nrf52_uart_nb)

puts with disabled interrupts and a full transmit buffer

Join us now and share the software;
You'll be free, hackers, you'll be free.
Join us now and share the software;
You'll be free, hackers, you'll be free.

Hoarders can get piles of money,
That is true, hackers, that is true.
But they cannot help their neighbors;
That's not good, hackers, that's not good.

When we have enough free software
At our call, hackers, at our call,
We'll kick out those dirty licenses
Ever more, hackers, ever more.

Join us now and share the software;
You'll be free, hackers, you'll be free.
Join us now and share the software;
You'll be free, hackers, you'll be free.

== printed in 2101951/2100000 µs ==
[SUCCESS]
```

### CODE REVIEW

File: cpu/nrf52/Makefile.features [hunk 1]
```diff
@@ -10,6 +10,9 @@ ifneq (,$(filter nrf52811xxaa nrf52820xxaa nrf52833xxaa nrf52840xxaa,$(CPU_MODEL
   FEATURES_PROVIDED += radio_nrf802154
 endif
 
+ifneq (,$(filter nrf5284%,$(CPU_MODEL)))
```
Comment (aabadie) [inline cpu/nrf52/Makefile.features:13]: AFAIK, only nrf52832 doesn't provide UARTE, so maybe just exclude this family ?
```suggestion
ifeq (,$(filter nrf52832%,$(CPU_MODEL)))
```

File: cpu/nrf5x_common/periph/uart.c [hunk 2]
```diff
@@ -221,33 +237,69 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
...
+                }
+                /* free one spot in buffer */
+                tx_buf[uart] = tsrb_get_one(&uart_tx_rb[uart]);
+                _write_buf(uart, &tx_buf[uart], 1);
+            }
+            tsrb_add_one(&uart_tx_rb[uart], data[i]);
+        }
+        else {
+            /* if no transmission is ongoing and ring buffer is full
+               free up a sport in the buffer by sending one byte */
```
Comment (aabadie) [inline cpu/nrf5x_common/periph/uart.c:286]: "free up a sport", what do you mean ?

Comment (fjmolinas) [reply cpu/nrf5x_common/periph/uart.c:286]: a spot*

File: cpu/nrf52/Makefile.features (hunk 1 above)
Comment (fjmolinas) [reply cpu/nrf52/Makefile.features:13]: It does not work on rnf51 either I believe.

Comment (aabadie) [reply cpu/nrf52/Makefile.features:13]: Yes, but this file is only for nrf52 :)

Comment (benpicco) [reply cpu/nrf52/Makefile.features:13]: it's really confusing that the nrf52832 is closer hardware wise to the nrf51 than the other members of the nrf52 family.
They should just have named it different.

Comment (fjmolinas) [reply cpu/nrf52/Makefile.features:13]: Anyway I prefer `ifneq` patterns than `ifeq` patterns if you don't mind. If there are other nrf52 susceptible to work I do not mind listing them.

Comment (aabadie) [reply cpu/nrf52/Makefile.features:13]: > Anyway I prefer ifneq patterns than ifeq patterns if you don't mind. If there are other nrf52 susceptible to work I do not mind listing them.

IMO, this is extra work each a new family is added, so subject to missing features unless someone remember to add it, and of course less maintainable.

**Comment by aabadie** [comment]
Features in Kconfig also needs an update :)

File: cpu/nrf52/Makefile.features [hunk 3]
```diff
@@ -10,7 +10,7 @@ ifneq (,$(filter nrf52811xxaa nrf52820xxaa nrf52833xxaa nrf52840xxaa,$(CPU_MODEL
   FEATURES_PROVIDED += radio_nrf802154
 endif
 
-ifneq (,$(filter nrf5284%,$(CPU_MODEL)))
+ifeq (,$(filter nrf5283%,$(CPU_MODEL)))
```
Comment (benpicco) [inline cpu/nrf52/Makefile.features:13]: nrf52833 should be like nrf52840, no?

Comment (fjmolinas) [reply cpu/nrf52/Makefile.features:13]: Did not know, will fix then

Comment (aabadie) [reply cpu/nrf52/Makefile.features]: I can test this on microbit-v2 (nrf52833 based) later today.

**Comment by fjmolinas** [comment]
Murdock is green.

File: cpu/nrf5x_common/periph/uart.c [hunk 4]
```diff
@@ -221,33 +237,69 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
         NRF_UART0->INTENSET = UART_INTENSET_RXDRDY_Msk;
         NRF_UART0->TASKS_STARTRX = 1;
 #endif
-
-        /* enable global and receiving interrupt */
+    }
+    if (rx_cb || IS_USED(MODULE_PERIPH_UART_NONBLOCKING)) {
         NVIC_EnableIRQ(UART_IRQN);
     }
-
```
Comment (benpicco) [inline cpu/nrf5x_common/periph/uart.c:228]: I'd leave the newline here

File: cpu/nrf5x_common/periph/uart.c [hunk 5]
```diff
@@ -221,33 +237,69 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
...
+    }
 }
 
 void uart_write(uart_t uart, const uint8_t *data, size_t len)
 {
     assert(uart < UART_NUMOF);
-
+#ifdef MODULE_PERIPH_UART_NONBLOCKING
+    for (size_t i = 0; i < len; i++) {
+        if (irq_is_in() || __get_PRIMASK()) {
```
Comment (benpicco) [inline cpu/nrf5x_common/periph/uart.c:271]:
```suggestion
        /* in IRQ or interrupts disabled */
        if (irq_is_in() || __get_PRIMASK()) {
```

File: cpu/nrf5x_common/periph/uart.c [hunk 6]
```diff
@@ -266,14 +318,15 @@ void uart_write(uart_t uart, const uint8_t *data, size_t len)
...
         _write_buf(uart, data, len);
     }
+#endif
 }
 
 void uart_poweron(uart_t uart)
 {
     assert(uart < UART_NUMOF);
 
     if (isr_ctx[uart].rx_cb) {
```
Comment (benpicco) [inline cpu/nrf5x_common/periph/uart.c:328]: Should this also be
```suggestion
    if (isr_ctx[uart].rx_cb | IS_USED(MODULE_PERIPH_UART_NONBLOCKING)) {
```
?

**Review by benpicco — COMMENTED** [review]
Looks good, just some small comments - feel free to squash directly

File: cpu/nrf5x_common/periph/uart.c (hunk 6 above)
Comment (fjmolinas) [reply cpu/nrf5x_common/periph/uart.c:328]: Nope, this is unrelated to the TX isr, which is the one impacted by the non-blocking functionality.

**Comment by fjmolinas** [comment]
May I squash @benpicco?

**Comment by benpicco** [comment]
Sure, please squash

**Comment by fjmolinas** [comment]
I may be missing something, I had an issue when using this in another application. I want to investigate where the issue comes from before.

**Comment by fjmolinas** [comment]
> I may be missing something, I had an issue when using this in another application. I want to investigate where the issue comes from before.

The issue is unrelated.

**Comment by fjmolinas** [comment]
Squashed

**Comment by fjmolinas** [comment]
@aabadie do you think you can test this one soon?

**Review by aabadie — APPROVED** [review]
Tested on microbit-v2 (nrf52833):

<details>

```
$ BUILD_IN_DOCKER=1 make BOARD=microbit-v2 -C tests/periph_uart_nonblocking/ flash test --no-print-directory 
Launching build container using image "riot/riotbuild:latest".
docker run --rm --tty --user $(id -u) -v '/usr/share/zoneinfo/Europe/Paris:/etc/localtime:ro' -v '/work/riot/RIOT:/data/riotbuild/riotbase:delegated' -e 'RIOTBASE=/data/riotbuild/riotbase' -e 'CCACHE_BASEDIR=/data/riotbuild/riotbase' -e 'BUILD_DIR=/data/riotbuild/riotbase/build' -e 'RIOTPROJECT=/data/riotbuild/riotbase' -e 'RIOTCPU=/data/riotbuild/riotbase/cpu' -e 'RIOTBOARD=/data/riotbuild/riotbase/boards' -e 'RIOTMAKE=/data/riotbuild/riotbase/makefiles'      -e 'BOARD=microbit-v2'  -w '/data/riotbuild/riotbase/tests/periph_uart_nonblocking/' 'riot/riotbuild:latest' make 'BOARD=microbit-v2'    
Building application "tests_periph_uart_nonblocking" for "microbit-v2" with MCU "nrf52".

"make" -C /data/riotbuild/riotbase/boards/microbit-v2
"make" -C /data/riotbuild/riotbase/core
"make" -C /data/riotbuild/riotbase/cpu/nrf52
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common
"make" -C /data/riotbuild/riotbase/cpu/cortexm_common/periph
"make" -C /data/riotbuild/riotbase/cpu/nrf52/periph
"make" -C /data/riotbuild/riotbase/cpu/nrf52/vectors
"make" -C /data/riotbuild/riotbase/cpu/nrf5x_common
"make" -C /data/riotbuild/riotbase/cpu/nrf5x_common/periph
"make" -C /data/riotbuild/riotbase/drivers
"make" -C /data/riotbuild/riotbase/drivers/periph_common
"make" -C /data/riotbuild/riotbase/sys
"make" -C /data/riotbuild/riotbase/sys/auto_init
"make" -C /data/riotbuild/riotbase/sys/div
"make" -C /data/riotbuild/riotbase/sys/isrpipe
"make" -C /data/riotbuild/riotbase/sys/malloc_thread_safe
"make" -C /data/riotbuild/riotbase/sys/newlib_syscalls_default
"make" -C /data/riotbuild/riotbase/sys/stdio_uart
"make" -C /data/riotbuild/riotbase/sys/test_utils/interactive_sync
"make" -C /data/riotbuild/riotbase/sys/tsrb
"make" -C /data/riotbuild/riotbase/sys/xtimer
   text	   data	    bss	    dec	    hex	filename
  12016	    128	   2524	  14668	   394c	/data/riotbuild/riotbase/tests/periph_uart_nonblocking/bin/microbit-v2/tests_periph_uart_nonblocking.elf
/work/riot/RIOT/dist/tools/openocd/openocd.sh flash /work/riot/RIOT/tests/periph_uart_nonblocking/bin/microbit-v2/tests_periph_uart_nonblocking.elf
### Flashing Target ###
Open On-Chip Debugger 0.10.0+dev-01406-gca211373d-dirty (2020-10-26-22:35)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
swd
Info : CMSIS-DAP: SWD  Supported
Info : CMSIS-DAP: FW Version = 0255
Info : CMSIS-DAP: Serial# = 9904360259304e450028901500000060000000009796990b
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : SWCLK/TCK = 1 SWDIO/TMS = 1 TDI = 0 TDO = 0 nTRST = 0 nRESET = 1
Info : CMSIS-DAP: Interface ready
Info : clock speed 1000 kHz
Info : SWD DPIDR 0x2ba01477
Info : nrf52.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : starting gdb server for nrf52.cpu on 0
Info : Listening on port 40187 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* nrf52.cpu          cortex_m   little nrf52.cpu          unknown

target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x00001a34 msp: 0x20000200
Info : nRF52833-xxAA(build code: A0) 512kB Flash, 128kB RAM
Warn : Flash protection of this nRF device is not supported
Warn : Adding extra erase range, 0x00002f70 .. 0x00002fff
auto erase enabled
wrote 12144 bytes from file /work/riot/RIOT/tests/periph_uart_nonblocking/bin/microbit-v2/tests_periph_uart_nonblocking.elf in 0.814142s (14.567 KiB/s)

verified 12144 bytes in 0.454861s (26.073 KiB/s)

shutdown command invoked
Done flashing
r
/work/riot/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" --no-reconnect --noprefix --no-repeat-command-on-empty-line 
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
READY
s
START
main(): This is RIOT! (Version: 2021.04-devel-725-ge2570-review_nrf_non_blocking)
                                                                
puts with disabled interrupts and a full transmit buffer

Join us now and share the software;
You'll be free, hackers, you'll be free.
Join us now and share the software;
You'll be free, hackers, you'll be free.

Hoarders can get piles of money,
That is true, hackers, that is true.
But they cannot help their neighbors;
That's not good, hackers, that's not good.

When we have enough free software
At our call, hackers, at our call,
We'll kick out those dirty licenses
Ever more, hackers, ever more.

Join us now and share the software;
You'll be free, hackers, you'll be free.
Join us now and share the software;
You'll be free, hackers, you'll be free.

== printed in 2101764/2100000 µs ==
[SUCCESS]
```

</details>

All good for me.

ACK!

**Pull request merged**